src/components/Chart/data/bitDataStorage.js

```diff
@@ -0,0 +1,111 @@
+ *
+ * 3. This software, with or without modification, must only be used with a Nordic
+ *    Semiconductor ASA integrated circuit.
+ *
+ * 4. Any software provided in binary form under this license must not be reverse
+ *    engineered, decompiled, modified and/or disassembled.
+ *
+ * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/* eslint no-plusplus: off */
+
+import { nbDigitalChannels } from '../../../globals';
+import { lineDataForBitState } from '../../../utils/bitConversion';
+
+const emptyArray = () =>
+    [...Array(4000)].map(() => ({ x: undefined, y: undefined }));
+
+export default () => ({
+    lineData: [...Array(nbDigitalChannels)].map(() => ({
+        mainLine: emptyArray(),
+        uncertaintyLine: emptyArray(),
+    })),
+    bitIndexes: new Array(nbDigitalChannels),
+    previousBitStates: new Array(nbDigitalChannels),
+
+    initialise(digitalChannelsToCompute) {
+        this.bitIndexes.fill(0);
+        this.previousBitStates.fill(null);
+        this.digitalChannelsToCompute = digitalChannelsToCompute;
+    },
+
+    storeEntry(timestamp, bitNumber, bitState) {
+        const current = this.lineData[bitNumber];
+        const bitIndex = this.bitIndexes[bitNumber];
+        const lineData = lineDataForBitState[bitState];
+
+        current.mainLine[bitIndex].x = timestamp;
+        current.mainLine[bitIndex].y = lineData.mainLine;
+
+        current.uncertaintyLine[bitIndex].x = timestamp;
+        current.uncertaintyLine[bitIndex].y = lineData.uncertaintyLine;
+
+        ++this.bitIndexes[bitNumber];
+    },
+
+    storeBit(timestamp, bitNumber, bitState) {
+        this.latestTimestamp = timestamp;
+
+        const bitChanged = this.previousBitStates[bitNumber] !== bitState;
+        if (bitChanged) {
+            this.storeEntry(timestamp, bitNumber, bitState);
+
+            this.previousBitStates[bitNumber] = bitState;
+        }
+    },
+
+    addFinalEntries() {
+        this.digitalChannelsToCompute.forEach(i => {
+            const hasEntry = this.bitIndexes[i] > 0;
+            const lastEntryIsNotForLastTimestamp =
+                this.latestTimestamp !==
+                this.lineData[i].mainLine[this.bitIndexes[i] - 1]?.timestamp;
```

Comment:
What? `?.` This looks very convenient ;)

Reply:
It is! :-) I was happy to discover in Dee’s code, that we are already able to use it.